### ISSUE TITLE
Backchannel URL rewriting skips port if hostname matches KC_HOSTNAME

### ISSUE THREAD

**Issue by jblachly** [keycloak/keycloak#16301]
### Before reporting an issue

- [X] I have searched existing issues
- [X] I have reproduced the issue with the latest release

### Area

core

### Describe the bug

Backchannel URLs are by default rewritten according to the incoming `Host` header. This can be optionally disabled with `--hostname-strict-backchannel=true`. By way of an example, retrieving the open-id configuration from `http://keycloak:8080/realms/vv-prod/.well-known/openid-configuration` with curl yields:

```
{"issuer":"https://auth/realms/vv-prod","authorization_endpoint":"https://auth/realms/vv-prod/protocol/openid-connect/auth","token_endpoint":"http://keycloak:8080/realms/vv-prod/protocol/openid-connect/token","introspection_endpoint":"http://keycloak:8080/realms/vv-prod/protocol/openid-connect/token/introspect","userinfo_endpoint":"http://keycloak:8080/realms/vv-prod/protocol/openid-connect/userinfo",...
```

when `KC_HOSTNAME=auth`. Note that some URLs are written as `https://auth/`, while backchannel-able URLs, like `token_endpoint`, are rewritten according to the `Host` header (`keycloak:8080`, in this case, since that is one of the hostname aliases on my docker network. **NOTE THE INCLUSION OF PORT NUMBER**).

The problem is that I had set my internal services to also use the `auth` name for backchannel, private-network requests, but when requesting URL `http://auth:8080/realms/vv-prod/.well-known/openid-configuration`, the port is stripped off (or more likely, backchannel-able URLs are just not rewritten at all):

```
{"issuer":"https://auth/realms/vv-prod","authorization_endpoint":"https://auth/realms/vv-prod/protocol/openid-connect/auth","token_endpoint":"https://auth/realms/vv-prod/protocol/openid-connect/token","introspection_endpoint":"https://auth/realms/vv-prod/protocol/openid-connect/token/introspect","userinfo_endpoint":"https://auth/realms/vv-prod/protocol/openid-connect/userinfo","end_session_endpoint":"https://auth/realms/vv-prod/protocol/openid-connect/logout","frontchannel_logout_session_supported":true,"frontchannel_logout_supported":true,"jwks_uri":...
```

This is problematic, because my internal services must access Keycloak through backchannel URLs, but keycloak is only listening on 8080/8443 on the internal network. In the meantime, I can set the internal network services to use a hostname alias.

(NOTE: for completeness, of course I am using proxy=edge mode)

### Version

20.0.0

### Expected behavior

I expect that when the `Host` header matches `<KC_HOSTNAME>:<port that's not 80 or 443>` for backchannel-able URLs to still be rewritten, just as they are when `Host` header matches `<anything other than KC_HOSTNAME>:<port>`

### Actual behavior

Backchannel URLs are not rewritten if the host-part, excluding port, of the `Host` header matches `KC_HOSTNAME`

### How to Reproduce?

Example docker-compose fragment included; the build `Dockerfile.example` is taken verbatim from the Keycloak documentation

```
    #image: quay.io/keycloak/keycloak:${KEYCLOAK_VERSION}
    build:
      context: keycloak
      dockerfile: Dockerfile.example
    restart: always
    healthcheck:
      test: ["CMD", "curl", "-f", "http://localhost:8080/health"]
      interval: 5s
      timeout: 2s
      retries: 15
    depends_on:
      - postgres
    environment:
      # Keycloak Docker-only options
      - KEYCLOAK_ADMIN=admin
      - KEYCLOAK_ADMIN_PASSWORD=admin
      # Keycloak Java options
      - KC_DB=postgres
      - KC_DB_URL=jdbc:postgresql://postgres/keycloak
      - KC_DB_USERNAME=postgres
      - KC_DB_PASSWORD=postgres
      - KC_HEALTH_ENABLED=true
      - KC_METRICS_ENABLED=true
      - KC_PROXY=edge
      - KC_HOSTNAME=${KEYCLOAK_HOSTNAME}
      - KC_HOSTNAME_STRICT_BACKCHANNEL=false
    command: start
    networks:
      backend-net:
        aliases:
          - auth
      traefik:
        aliases:
          - auth
```

**Comment by bbednarek** [comment]
I have exactly the same problem. Please check https://github.com/keycloak/keycloak/discussions/16482 for more details.

**Comment by mposolda** [comment]
Thanks for the report, but unfortunately due the amount of other reported issues and other priorities, Keycloak team does not have time to properly triage this bug. So preliminary added to Backlog for now.
It will be helpful if:
- You can verify if still applicable in latest Keycloak released version. If not, then it is welcome to close this issue.
- If you figure that this may not be a valid bug (for example just a mistake in configuration etc), it will be also welcome to close this issue
- If still applicable in latest version, it will be welcome to add the comment as well, that this was still reproduced with latest Keycloak version as it is very valuable info. Anyone is welcome to comment with this or add other relevant comments to this issue.

**Comment by bdabelow** [comment]
I think you also need to set hostname-port or use hostname-url.

**Comment by vmuzikar** [comment]
@jblachly I'm sorry for the very late reply. This looks like a real issue due to:
https://github.com/keycloak/keycloak/blob/74b56201c32e0b7b8b62ee6b0576e502392018e1/quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/hostname/DefaultHostnameProvider.java#L159-L161

Removing the triage label but it will be fixed with the Hostname V2: #24252.

**Comment by vmuzikar** [comment]
This will be resolved by the new hostname provider.

**Comment by vmuzikar** [comment]
It will be resolved with #24252.

**Comment by vmuzikar** [comment]
This issue has been created against Hostname v1 implementation. Due to its complexities and high maintenance costs we decided to rewrite it from scratch as Hostname v2 which will be part of Keycloak 25 and will be replacing v1. For more context please see #24252.

We believe this issue has been resolved with Hostname v2. We encourage you to give it a try even before Keycloak 25 is released and report back any issues you run into. You can use the [nightly builds](https://github.com/keycloak/keycloak/releases/tag/nightly) for it. See also the [migration notes](https://github.com/keycloak/keycloak/blob/main/docs/documentation/upgrading/topics/changes/changes-25_0_0.adoc#new-hostname-options) and [new docs](https://github.com/keycloak/keycloak/blob/e7dd5c1991e938ad2aa03e9c8b18cac6820dc7df/docs/guides/server/hostname.adoc).
